### PR TITLE
feat: weather constraint evaluation

### LINKED ISSUE #27: Enforce weather constraints (TWS < 20 m/s, Hs < 7 m)
## Context
SWOPP3 specifies two weather constraints:
- True wind speed < 20 m/s
- Significant wave height < 7 m

If it is not possible to keep the limit, disregard the limits. RISE will register to what extent the limits were kept.

## Requirements
- Add constraint penalties or hard barriers in the cost function for:
  - TWS exceeding 20 m/s at any waypoint along the route
  - Hs exceeding 7 m at any waypoint along the route
- Report max TWS and max Hs per voyage in the output (required columns in File A)
- Configurable constraint thresholds

## Dependencies
- #25 (ERA5 data for wind speed and wave height fields)

## Acceptance Criteria
- [ ] Cost function augmented with weather constraint penalties
- [ ] Max wind and max Hs tracked per route evaluation
- [ ] Constraints can be relaxed when infeasible
- [ ] Unit tests for constraint enforcement

### PR DESCRIPTION
**Pull request by fjsuarez**
## Summary
Add weather safety constraint checking for route optimization.

### New files
- `routetools/weather.py`: `weather_penalty()` evaluates TWS and Hs along routes, returning penalty proportional to segments exceeding configurable limits. JIT-compatible.
- `tests/test_weather.py`: Tests for calm conditions, exceeding limits, combined penalties, edge cases.

### Dependencies
None (independent)

---
*Part 4 of 9 — extracted from vangelis branch for clean review*

---
Closes #27

### CODE REVIEW

File: routetools/weather.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    -------
+    jnp.ndarray
+        Penalty per route, shape ``(B,)``.
+    """
+    mid_lon, mid_lat, t_zeros = _segment_midpoints(curve)
+
+    violations = jnp.zeros(curve.shape[0])
+
+    if windfield is not None:
+        u10, v10 = windfield(mid_lon, mid_lat, t_zeros)
```
Comment (daniprec) [inline routetools/weather.py]: Why `t_zeros`? The wind and waves change daily (I think hourly in the case of waves, maybe wind too). Are we sure we are using the appropriate timestamps?

**Review by daniprec — CHANGES_REQUESTED** [review]
## PR Review — Blockers & Should Fix

### Overall

The approach looks reasonable. A simple safety penalty that stays JIT-compatible fits the routing loop well. The tests listed cover basic behavior.

I have one major concern about time handling.

---

### Blockers

1. Weather is evaluated at t=0 for all segments

You sample spatial midpoints along the route, which is fine. But you set times to zeros, so every segment uses the same timestamp.

That breaks the core assumption of weather routing. Wind and waves vary with time. If you evaluate everything at a single time, you can mark an unsafe segment as safe, or the reverse.

Why are we using `t_zeros`? Are we sure the weather fields are time-aware and that we pass the correct timestamps per segment?

At minimum, the penalty needs to evaluate weather at the segment timestamp, based on departure time plus cumulative travel time along the route.

---

### Should Fix

2. Add a time-variation test

Add a test with a synthetic time-varying field to prove timestamps are used correctly.

Example setup:

* Route with at least 3 segments
* Segment times: day 0, day 1, day 2
* Weather field:

  * day 0 below limit
  * day 1 above limit
  * day 2 below limit

Expected result:

* Penalty is non-zero only due to the day 1 segment
* If the implementation still uses t=0 everywhere, the test will fail

This test will catch the exact bug risk here and will protect future refactors.

---

### Recommendation

Do not merge until the time handling is corrected or clearly justified, and the time-variation test passes.

**Comment by fjsuarez** [comment]
Good catch — you're right that the weather constraint evaluation uses `t=0` for all segments. This is a real bug.

### 1. `t_zeros` — why it exists and why it's wrong

The `_segment_midpoints` helper hard-codes `t_zeros = jnp.zeros_like(mid_lon)`, which is then passed to `windfield(mid_lon, mid_lat, t_zeros)` and `wavefield(mid_lon, mid_lat, t_zeros)`. Every segment evaluates weather at departure time regardless of where it falls along the route.

The cost function in `cost.py` already handles this correctly — `cost_function_constant_speed_time_variant` uses `lax.scan` to accumulate `t + dt` at each step and passes the running `t` to the vectorfield. But `weather.py` doesn't reuse that logic.

### Fix plan

The fix is to compute per-segment elapsed times from the curve geometry and ship speed, then pass those as the `t` argument instead of zeros:

```python
def _segment_midpoints(curve, travel_stw, spherical_correction=False):
    mid_lon = (curve[:, :-1, 0] + curve[:, 1:, 0]) / 2
    mid_lat = (curve[:, :-1, 1] + curve[:, 1:, 1]) / 2
    # Compute segment lengths and cumulative travel time
    if spherical_correction:
        dx, dy = haversine_meters_components(...)
    else:
        dx = jnp.diff(curve[:, :, 0], axis=1)
        dy = jnp.diff(curve[:, :, 1], axis=1)
    segment_dist = jnp.sqrt(dx**2 + dy**2)
    segment_dt = segment_dist / travel_stw  # seconds per segment
    # Cumulative time at each segment midpoint
    t_cumulative = jnp.cumsum(segment_dt, axis=1) - segment_dt / 2
    return mid_lon, mid_lat, t_cumulative
```

This is a first-order approximation (it ignores the wind/current effect on ground speed), but it's far better than `t=0` everywhere and avoids the circular dependency of needing weather to compute time to compute weather.

### 2. Time-variation test

Agreed — will add a synthetic test with a 3-day field where only day 1 exceeds the limit, as you described. This will catch regressions if someone reverts to constant-time evaluation.

**Review by daniprec — APPROVED** [review]
Looks good to me! Merge.
